pilot/pkg/networking/core/v1alpha3/listener.go

```diff
@@ -451,6 +467,16 @@ func (lb *ListenerBuilder) buildSidecarOutboundListeners(node *model.Proxy,
 
 					// bind might have been modified by below code, so reset it for every Service.
 					listenerOpts.bind = bind
+					// bind and extraBind should be specified at the same time for dual stack env
+					// and make sure that it's necessary to add the extra addresses, such as it's a dual stack env
+					if actualLocalHostAddrIPv6 != "" {
+						if bind == actualLocalHostAddrIPv4 {
+							listenerOpts.extraBind = []string{actualLocalHostAddrIPv6}
```

Comment:
do we want to make `bind` and `extraBind`? Should we just have `bind` be a list and we automatically make them "extra"?

In envoy, is there any difference between the primary address and the "extra" addresses?

Reply:
> do we want to make `bind` and `extraBind`? Should we just have `bind` be a list and we automatically make them "extra"?
> 
> In envoy, is there any difference between the primary address and the "extra" addresses?

Hi @howardjohn, as discussing in previous PR, `ClusterVIPs` can holds all IP instances which is available both for `bind` and `extraBind`.

Yes, according to the issue https://github.com/envoyproxy/envoy/issues/11184 in envoy, `additinal_addresses` can hold all other IP instances for one specified listener, except the primary one. And [here ](https://github.com/envoyproxy/envoy/pull/21391/files)is the API definition for it.

Reply:
@howardjohn  For example, if there are 2 IP instances ["127.0.0.1", "::1"],  then the primary address should be `127.0.0.1`, and the extra addresses should be [`::1`]

Reply:
I get that's how it's implemented but my question is why? Do we really need to carry around a "address" and "extra addresses" or just a list of addresses and stick a random one as the main on in envoy config?

My concern with this PR, as noted before, is not the generates XDS but the large complexity we are adding to the code for something relatively simple, and lack of integration tests.

If we need all of these complex if statements and juggling around 4 address (localhost + wildcard x v4 v6) it's complex and error prone. Maybe this logic can be factored out into a struct, etc

Reply:
Hi @howardjohn, how about to factor the extra addresses adding logic into the function of structs `buildListenerOpts` and `inboundChainConfig`, then the similar if statements, such as `if len(cc.extraBind) > 0 && util.IsIstioVersionGE116(lb.node.IstioVersion)` will be eliminated.
For integration test, the dual stack integration test will be added in later and the use case would be based on the really user scenarios. @kfaseela, @zhanghw0354 and I would co-work on it.

Reply:
We don't need to make all of our code much more complex just because envoy has a weird API, imo. We should pick the right API for our usage and then translate it to envoy as needed

Reply:
Integration tests are an expectation for all features in Istio; please have them alongside the PR instead of afterwards. We can wait. If you want to merge without integration tests it can be done in the experimental branch

Reply:
> We don't need to make all of our code much more complex just because envoy has a weird API, imo. We should pick the right API for our usage and then translate it to envoy as needed

Sure,  agree with this idea, I will make the code as simple as possible, then re-submit it.

Reply:
> Integration tests are an expectation for all features in Istio; please have them alongside the PR instead of afterwards. We can wait. If you want to merge without integration tests it can be done in the experimental branch

@howardjohn  Sure, I will add integration tests for this feature, and make sure the `additional_addresses` can be correctly added. Thanks for the comment here.